### ISSUE TITLE
Add async constructor for Materials with image textures

### ISSUE THREAD

**Issue by jjhembd** [CesiumGS/cesium#10566]
A [Material](https://cesium.com/learn/cesiumjs/ref-doc/Material.html) instance currently gives no indication of whether its textures are loaded or not. This makes unit testing difficult. 

For example, the following test will fail since the image is not loaded before the scene renders:
```javascript
it("renders blue from a blue texture", function () {
  primitive.appearance.material = Material.fromType("Image", { 
    image: "./Data/Images/Blue.png"
  });
  scene.primitives.add(primitive);
  // ... Set camera view, etc...

  expect(scene).toRender([0, 0, 255, 255]);
  // Error: Expected to render [0,0,255,255], but actually rendered [255,255,255,255].
});
```

Some [current tests](https://github.com/CesiumGS/cesium/blob/dc28421ff2f52871629299d00f310ee796f63833/Specs/Scene/MaterialSpec.js#L634-L643) fix the problem by waiting for a change in the length of the `material._loadedImages` array, but this is complicated to set up.

Proposal: add a static method `Material.fromUrl` returning a Promise. The Promise would resolve to a new Material instance with all textures ready for use. The above test would become:
```javascript
it("renders blue from a blue texture", async function () {
  const material = await Material.fromUrl({
    type: "Image",
    image: "./Data/Images/Blue.png",
   });
  primitive.appearance.material = material;
  scene.primitives.add(primitive);
  // ... Set camera view, etc...

  expect(scene).toRender([0, 0, 255, 255]);
});
```

**Comment by gaurav8y785675474** [comment]
hello jjhembd
i am a beginner in open source world
can you please tell me exact location of code[related to this issue] in files
so that i can work on that

**Comment by ggetz** [comment]
Hello @gaurav8y785675474,

The new function should be added to `packages/engine/Source/Scene/Material.js`.

Then search the unit tests for occurrences of `material._loadedImages`. That should be replaces with the async/await pattern shown above.

**Comment by wesleykebrown** [comment]
Hi, I’m a part of the Flagship program for JTC. I would like to work on this issue ticket, please! This is a great opportunity to test my abilities.

**Comment by ggetz** [comment]
@wesleykebrown Sounds great!

**Comment by adamwirth** [comment]
I've been working on this for a while locally and wanted to ask some questions!

Questions I have for solution paths:

1. Could the tests be modified to use `pollToPromise`?
1. With the image urls, there's a nested call in generated functions from `createTexture2DUpdateFunction` ([here](https://github.com/CesiumGS/cesium/blob/main/packages/engine/Source/Scene/Material.js#L787C10-L787C39)]. When interacting with `_loadedImages`, this function's path can either be synchronous or asynchronous. The asynchronous path has potentially multiple `.then` calls, which is probably what's causing the test issue. For improving this, should this and its parent functions be refactored to have a split async/sync path depending on if there's a "url" passed as an arg to a constructor? This would be a lot of refactoring.
1. The variant of above is making every function async and returning the promise chains but that's even worse.

<details>
<summary>
Questions I have for code understanding in this starbucks (hidden cuz its like, optional and I'm fun)
</summary>

1. I've been reading the [coding guide](https://github.com/CesiumGS/cesium/blob/main/Documentation/Contributors/CodingGuide/README.md) & I wanted to ask where this `Material` falls in the range of low-level to high-level
1. From the coding guide, it suggests writing code similar to what already exists. Unfortunately, for the `fromUrl` results in the repo, they're wildly different implementations. Mine more closely resembles [the more recently updated one of them](https://github.com/CesiumGS/cesium/blob/main/packages/engine/Source/Scene/SingleTileImageryProvider.js), but it is postfixed with `Provider`, and I wasn't sure if that had any implications
1. I saw in the op the example fix has a parameter structure breaking with the `fromUrl` tradition: having `url` be the first param. Would it be ok to have `url` be the first param or was that intended as its considered `image` as a nested property within the `Material` spec?
1. Also looking at other implementations of fromUrl, they're all public. Since the intention here is for improving testing, would it make sense to break from that here and have it be private or no?
</details>

Thanks

**Comment by bkuster** [comment]
we are actually facing a similar challenge, but in a custom material using fabrics. would your solution also work for this?

i hacked a quick prototype where i used an `imageLoaded` event that would emit on `update` for every loaded image with the uniformId. the benefit of this, i thought, was a) that i can change the uniform and wait for the new texture to be loaded again without having to create a new material and b) the code changes where minimal: a total of three lines. BUT, you have to actually call update on the material. 

maybe there is an inbetween solution that could take care of both the `await Material.fromUrl('my.png')` and `new Material({ fabric: { type: 'MyFabric', uniforms: { tex: myCanvas }, source: '' } })`.

**Comment by mzschwartz5** [comment]
If the only purpose of `Material.fromUrl` is to make unit testing easier, why don't we instead make some sort of `imageReady` flag and event that Material creators can listen for? Unit tests could enter a poll loop on the ready flag, or register a listener which performs the `expects` validation steps.

I could see it being useful in other places in the codebase to have a flag that indicates an image is loaded. On the other hand - do we have use cases where we'd want to asynchronously load an image from a URL?

**Comment by javagl** [comment]
This sounds related to (although not "the same as") the changes that had been done in https://github.com/CesiumGS/cesium/blob/main/CHANGES.md#1107---2023-07-03 . There, the move was generally to change existing cases of "ready flags" and events into actual promises, but no strong opinion on the `Material.fromUrl` case from my side.

**Comment by mzschwartz5** [comment]
I see. In that case, maybe we should just store an array of promises (of images to load) in the material class, and expose a public getter so that users can await those promises? Users can continue to make materials as before - via the constructor or `.fromType` - but can subsequently await the images loading.

We could also then allow users to (optionally) pass in functions to call on resolving the promise, etc.

**Comment by javagl** [comment]
I might be lacking some context for the Material API in particular. It sounds like the suggested `Material.fromUrl` function is not _necessarily_ intended _only_ for tests, but could also be useful in other contexts. Conversely, people might want to have finer control, to optionally wait for (and/or be informed about) the image loading process, but I might be lacking the imagination about why they could want this. For the example of
```
  primitive.appearance.material = Material.fromType("Image", { 
    image: "./Data/Images/Blue.png"
  });
```
it looks like it can hardly make sense to use this material _without_ the image (this could cause... you know... some white flashing and such 😬 ). But I think that images can also be "composed" and contain _multiple_ external images, and being able to track this with dedicated promises for each image sounds pretty involved (for the implementation, _and_ for the client who wants to use it).

From the description until now, it sounds like a "blanket"-function like `fromType`**`Async`** could make sense: It creates whatever has to be created, returns a promise, and when it's done, it's done. But maybe there are use-cases for the "fine-grained" promises as well.

**Comment by mzschwartz5** [comment]
> dedicated promises for each image sounds pretty involved

Is it? If the `Material` class holds an array of image promises that's gettable, couldn't the user simply `Promise.all(theArray)` to wait for all of them?

**Comment by mzschwartz5** [comment]
I'm not necessarily arguing against an async constructor, for what it's worth. Just playing devil's advocate to try to understand *exactly* what we want here, before wasting effort building something we don't.

I do think something like `.fromTypeAsync` would be nice to have as part of the Material interface. It *will* be a larger refactor effort, though (an overdue one, to be fair).

**Comment by javagl** [comment]
> couldn't the user simply `Promise.all(theArray)` to wait for all of them?

Sure, but then I'd ... (also play devil's advocate, and) ... ask a few questions 🙂 

From the user's perspective, when the user wants to wait, as in "Just give me that material!", then there are these options:
```javascript
const material = await Material.fromTypeAsync(type, { ... });
```
vs.
```javascript
const material = Material.fromType(type, { ... });

// Important: Wait for it!
await Promise.all(material.getImagePromises());
```
It's a tad less convenient, and new users could easily miss that call. 

And... does the material even _contain_ images? Maybe, maybe not. That array of promises might be empty...

Also, offering that array of promises is exposing something that - to my very limited understanding of how that API is commonly used  - could very well remain hidden, unless there is a _strong_ reason why a user should _explicitly_ be able to do something like
`material.getImagePromises()[42].then(...);`
(And I can hardly imagine why a user should want to do that - maybe I'm overlooking something...)

(The case where a user does something like `material.getImagePromises()[42] = undefined;` may look contrived, but for the sake of "information hiding" etc., it may be worth mentioning that this would be _possible_)

Even _if_ this array is accessible: The user does not even _know_ what each element of this array represents. I'd have to dig through https://github.com/CesiumGS/cesium/wiki/Fabric and try out a few things, but from my understanding there can be materials that involve multiple images...
```
{ 
    imageA: "./Data/Images/Red.png"
....
    imageB: "./Data/Images/Blue.png"
}
```
Will the array of promises contain the promises of each of them, in declaration order? 

---

Note that I'm also not (strongly) arguing for or against one thing, except for the "high-level" points:  

1. the API should be _easy to use_, and _hard to misuse_
2. when in doubt, leave it out

**Comment by mzschwartz5** [comment]
👍🏻 I buy it :) Will update this thread when I have a better understanding of the refactoring that would be necessary to make an async material constructor.

**Comment by javagl** [comment]
Don't spend time and effort until this is confirmed. I'm not making decisions. I'm not arguing. I'm just trying to lay out pros and cons.

**Comment by mzschwartz5** [comment]
Just some thoughts as I'm going through the Material API (will probably edit this comment as I have more):
- A material can have submaterials. If a material is created via `fromTypeAsync`, I think that implies the submaterials would have to be handled similarly... recursively. That's pretty challenging, as submaterials are created from the Material constructor itself, which cannot be marked async.
- The solution here needs to continue to allow users to change material textures. 
- The solution should make it easy to add other asynchronous operations to `Material.js` that need to be waited on in an async constructor, in the future.
- Would be nice if the solution did not duplicate code / might require refactoring out loading logic into more easily callable functions.
- Is the solution acceptably "async" if it requires that the `Material` have an `Update` call before being truly "ready"?
- What do we do about `HTMLVideoElement`, `OffscreenCanvas`, etc. that aren't explicitly fetched in Material.js?

**Comment by mzschwartz5** [comment]
So, this is more challenging than I anticipated, but not impossible. It just requires making some design choices that I'd like input on.

Let me lay out the challenge first. There are currently two operations that are asynchronous in `Material` - loading images for 2D textures, and for cubemaps. They both occur via functors that are called from the `Update` loop. For example [this function](https://github.com/CesiumGS/cesium/blob/668c2050e19756a0effbd0ddaa613a350dbf30b3/packages/engine/Source/Scene/Material.js#L787-L789) creates a function which is called every Update to potentially load new images, if uniforms have changed.

So, the asynchronous parts of Material aren't even kicked off on construction - they're kicked off on the first update. This makes it hard to extract them into async functions that can be called in an async constructor.

Here are three approaches I've considered:
1. We extract the image loading logic into async functions, and call those explicitly in a new async constructor. 
2. We create a `ready` flag and event, which gets set / fires when everything is loaded the first time. The async constructor returns a promise which resolves on this event. To the user, it just looks like a regular promise that can be `await`'ed.
3. We mark `initializeMaterial` (and downstream calls all the way to `createUniform`) as `async`, and after [creating each updater function](https://github.com/CesiumGS/cesium/blob/668c2050e19756a0effbd0ddaa613a350dbf30b3/packages/engine/Source/Scene/Material.js#L1061), we immediately call and `await` it. The constructor would *not* `await` the `initializeMaterial` function (it can't since it's a constructor, but it shouldn't anyway). Instead, it stores the promise as a member variable. The new async constructor can await this promise.

The first approach sounds appealing because it doesn't require a ready flag/event pattern, but the logic in the uniform updaters (like `createTexture2DUpdateFunction`) is so complex that it would be really hard to do without any regressions. It's also not very future-proof; adding new async ops to Material would require adding them to the async constructor as well. I briefly tried this approach but got the ick as the rabbit hole of refactoring went ever deeper.

The second approach is honestly not bad, imo. It's probably the simplest in terms of refactoring, but ready flags are apparently an anti-pattern we're trying to move away from.

I think the third approach is the cleanest, and I tried doing something in this vein until I realized that one of the updater's requires `context`, which is only available in `Update` - ugh! If it's possible to extract the (limited) need for `context`, this is what I'd want to do. [Here's the only use that's really blocking](https://github.com/CesiumGS/cesium/blob/668c2050e19756a0effbd0ddaa613a350dbf30b3/packages/engine/Source/Scene/Material.js#L817-L820).

**Comment by javagl** [comment]
I'm like some of the `Material` functionality that is supposed to be pulled out: I'm lacking some `context` 🤡 

That context issue sounds pretty fundamental, and for me, raises the question about how _any_ of these three approaches should work.

It sounds like the material could only become `ready` when the images are loaded and the textures are created. And it sounds like this _requires_ the `update` function to be called. If this is correct, then I wonder: When someone does
`const material = await Material.createAsync(...);`
in a sandcastle (and nothing else), then there is no time and place where `update` is called at all, and it would wait forever (?). I think that _requring_ the `context` in any context where something `async` is supposed to happen brings us between a rock and a hard place: "I'd like to render (and call `update`), but for that, I need the material - which is waiting for the `update` call to happen". (Am I overlooking something obvious here...?)

The point about the changing uniforms: 

Even if there was that magic
`const material = await Material.createAsync(...);`
function that worked perfectly, one could ask what should happen when
`material.thatUniform = "newImage.png";`
is called. It has to wait for that PNG to be loaded, and then create the texture (using the `context`), and there is no time and place where one could "wait". The only solution for that is to do what is already done for the "white flashing" case - namely, keep what is there, until the "new" things is ready, and then switch it. 

Combining both points: There seem to be these two use-cases:

1. load at construction time (and resolve a promise when it's done)
2. load when a uniform changes (and switch out some 'texture' when it's done)

It might be possible to structure this in a way so that both paths use largely the same code. But that would require a clean separation of concerns. I've seen that (single!) 150-line-"updater" function that, as usual, is documented with
```

```
and that suggest that this might not be so easy So while separating "whatever should happen async" and "whatever needs the context" sounds like the only viable path here, I'm aware that there are unforeseeable refactoring efforts hidden in that high-level statement. 

(I probably have to take a closer look at the actual code here...)

**Comment by bkuster** [comment]
having haphazardly solved this for our custom material (as seen [here](https://github.com/virtualcitySYSTEMS/map-core/blob/c0dbbd41c1fe04587254467cbb69f9abd4986362/src/panorama/panoramaTileMaterial.ts#L298) where  i basically hack the update function to do that which i couldn't using just the Material API) i would like to add some "insights".

1. the context is needed to create the texture. it is _not_ however needed to load / create / manipulate the image data and bring it into a form where you can actually pass it to webgl. to get rid of the "white flashing" bit (which was my main use case, the other being that i only want to do a certain bit of my shader, if an image is even set), waiting on the image data to be loaded and setting a uniform to `true` once it is (in my case, lacking API, in my custom update function) was enough.
2. the major problem i see with the `.createAsync` approach, is that it assumes a very strict life cycle: create, view, dispose. as @javagl points out, this is not the case. i can change a uniform when ever. to me, this would call for some kind of event API, something a long the lines of `imageStateChanged`, which gets passed the uniform key and the state of the image `loading`, `failed` or `ready`. this i could then use as follows:
```js
// create my custom image
const material = new Material({ 
  fabric: { 
    type: 'MyMultiImage', 
    uniforms: { u_imageA: './a.png', u_ImageB: './b.png', u_aReady: false, u_bReady: false },
  },
});

// handle image ready state here. 
material.imageStateChanged.addEventListener((uniformKey, state) => {
    const  ready = state === 'ready';
    if (uniformKey === 'u_imageA') {
      material.uniforms.u_aReady = ready;
   } else if (uniformKey === 'u_ImageB') {
      material.uniforms.u_bReady = ready;
   }
});

// oh, something changed my texutre. no worries, my event handler
// will take care to change the ready flag to false and wait for the new image to be loaded
setTimeout(() => {
  material.uniforms.u_imageA = './other_a.png';
}, 2000)
```
those are my two cents, hope this helps further your designs.

PS my approach would need some tweaking to handle the Material.DefaultImageId which i have been using as an `undefined` value.

**Comment by mzschwartz5** [comment]
To @bkuster's comments:

> the context is needed to create the texture. it is not however needed to load / create / manipulate the image data

This is the "issue" I ran into above. The context is really only needed for the texture; it may be possible to do a refactor where my 3rd approach above works, if not for the `new Texture(...context...);` piece.

> this would call for some kind of event API, something a long the lines of `imageStateChanged`

I'm not entirely against this idea. It might be right for this situation. BUT, if we ever add other asynchronous operations to the Material initialization, this would no longer be sufficient. And what about for cubemap-based materials, where there are *6* images to load? 

Also, with a [recent fix](https://github.com/CesiumGS/cesium/pull/12722), Materials shouldn't flash white anymore when changing properties of an existing material. Only when doing an initial load; so a `fromTypeAsync` that awaits the initial load ought to cover our needs. (?)

To @javagl: 

> There seem to be these two use-cases...

I'm not so sure, honestly. Is there a use case for asynchronous uniform switching? Or is the load-and-switch sufficient for that? For new Materials, it's certainly not sufficient. For one thing, we've already identified that we want a truly asynchronous constructor for better unit testing, right?

**Comment by mzschwartz5** [comment]
I think you're right, @javagl, that my second approach would create an infinite loop of waiting for an Update that never happens... I thought maybe the render loop happens in a separate thread, but looks like not?

**Comment by mzschwartz5** [comment]
If [the HTMLVideoElement block](https://github.com/CesiumGS/cesium/blob/668c2050e19756a0effbd0ddaa613a350dbf30b3/packages/engine/Source/Scene/Material.js#L801-L833) is extracted to its own updater function, then nothing else will be dependent on `context` in the other updaters, and approach 3 might work. Going to play around a bit with this.

(Of course, still open to the idea that maybe we don't want an `async` constructor at all, but instead an `imagesReady` event or something like that)

**Comment by javagl** [comment]
> Is there a use case for asynchronous uniform switching?

Maybe that was a misunderstanding. From what I see, uniform switching is always "asynchronous", When the uniform is changed, the object/material already is displayed, with its "initial state". And I could not imagine what a user might want to "wait for" here after switching the uniform. 

Sure, the user might be _informed_ (with an event) when the new image is loaded, but that's the broader point: 

That `material.imageStateChanged.addEventListener` that was suggested there right now is something that focussed on an _internal_ aspect, namely setting that other uniform (to avoid some sort of flickering). 

But _if_ such a mechanism was in place, then it could also be used for handling asynchronicity. ("Handling" as in ~"making it synchronous, await-able"). This is what my "use-cases" referred to, and that _"...both paths use largely the same code"_. 

If we had this magic listener, then this could probably be used to avoid the flickering (internally), with pseudocode
```javascript
this.imageStateChanged.addEventListener((....) => {
  if (image is ready) { 
    // "Atomically" swap the internal structures - this is roughly what was done
    // with that swapping of the "oldPrimitive" and the new primitive, mentioned
    // in https://github.com/CesiumGS/cesium/issues/1640#issuecomment-3024929262
    discard(old image texture);
    use(new image texture); 
  }
});
```
And the same could be used for that async contruction (pesudocode again):
```javascript
Material.fromAsync() { 
  const material = create();
  const promise = pending(material);
  material.this.imageStateChanged.addEventListener((...) => {
    if (everything ready) {
      promise.resolve();    
    }
  });
  return promise;
});
```

(This could, in theory, even be applied for some `setUniformAsync` that returns a promise that resolves when the newly set image was loaded, but as I said: There may not be much use for that. At this point, one could rather consider "pulling that out", and offering some `material.setUniform("image.png", theActualImageThatWasAlreadyLoaded);` or so, but maybe that's a bit of a stretch)...)

I think that one of the most important aspects for all this is trying to more cleanly separate 1. the image loading and 2. "the part that needs the `context`", but can imagine that this could be a considerable effort. (I'm advocating for a hint in the coding guide, that no function should have more than 50 lines of code, just to avoid these ominous 150-line-`update` functions that are "doing various things". I'm not sure how easy it is to even isolate that part that needs the context, given corner cases like cubemaps, videos, and whatnot...)

**Comment by mzschwartz5** [comment]
> I'm advocating for a hint in the coding guide, that no function should have more than 50 lines of code

+100000. How about like 30 lol. Or at least single-purpose. And no nested conditionals more than 2 layers deep. Or super-duper compound conditionals.

**Comment by mzschwartz5** [comment]
@javagl 

```JS
Material.fromAsync() { 
  const material = create();
  const promise = pending(material);
  material.this.imageStateChanged.addEventListener((...) => {
    if (everything ready) {
      promise.resolve();    
    }
  });
  return promise;
});
```

This is what I was sort of what I was suggesting in my 2nd bullet in [this comment](https://github.com/CesiumGS/cesium/issues/10566#issuecomment-3111564508) but I think it causes an infinite hang? Because if a user awaits this, no Update ever gets called, so the image is never ready.

They could subscribe to the `imageStateChanged` event though - or, better yet, subscribe to a `materialReady` event.

**Comment by javagl** [comment]
This sounds like it might be related to your earlier [comment, about the material waiting for an update](https://github.com/CesiumGS/cesium/issues/10566#issuecomment-3113518129). But just _loading_ the image (i.e. sending out a request and being informed about the response) should be independent of the `update` calls. 

I just hacked together a toy example in NodeJS with a `fetch` for an image that triggers an event when the response arrives, and turning that into a promise to ensure that the (dummy) Material is ready, and that seemed to work. But this does not mean that there may not be quirks in the material or all the resource handling that cause this to ~"not work as expected". I'd have to dig into the code to see what could be wrong there, but is there anything happening in the `update` that could depend on this? E.g. could it be that the image requests are _only sent out(!)_ during the `update` calls? If they are sent out ~"from the constructor", then I think that this should work...

**Comment by mzschwartz5** [comment]
> But just loading the image (i.e. sending out a request and being informed about the response) should be independent of the update calls

Unfortunately not, without some refactoring. The request to load an image gets kicked off from the Update loop

> E.g. could it be that the image requests are only sent out(!) during the update calls? 

Yeah, that :)
